### ISSUE TITLE
Unordered ToDos 3 (Evaluation 2)

### ISSUE THREAD

**Issue by matrikular** [joomla-projects/GSoC17_publishing_workflow#42]
This is a list of notes as a result of a black box text (no code has been reviewed on my part, yet). Furthermore, I've tried my best to merge the tasks and notes from the unordered lists into this one.
It is very likely that some of the tasks are already done and that some of the mentioned bugs have already been fixed. For the next few "days", I would like to suggest that we look into the workflows MVC structure and try to get that ready.

Workflow(s):
- Add search / filter to list view
	Hint: Either global layout or custom layout depending on flexibility
		Possible Fields: [Search][Clear][Search Tools*][Status][Default (RFC / TBD][Author]
		Note: The ordering of those fields should reflect the order of the table headings (see other core components)
	Dependencies: Model, View (View Layouts), Global Layouts
	There are at least two exsting issues / tickets regarding the Search Tools Layout. One is about a bug that already has an influence on the
	list views "Clear" button for the search field. The second ist about a refactoring of the Search Tools in general.
	Attention: Both could become a blocker for this task! The same goes for all other filters including: Transitions and States!
- Add table ordering to list view. The default ordering should be: "Title ascending" (current is ID ascending)
	- Possible Sortable Headings: Status, Title, Created, Author, ID (RFC / TBD: Access)
		Note: The wording AND ordering should be the same across all extensions e.g. created fields (by vs at vs date)
- Add publish, unpublish and trash toggle functionality the same way it is implemented in other core extensions
- Add the following toolbar buttons to the list view: New, Edit, Publish, Unpublish, Default, Check-in, Trash
	None: Some fields change in wording and function depending on the users access level and / or the current entity state (e.g. Check-in, Trash)
- Add the following toolbar buttons to the add / edit form view: Save / Save & Close / Save & New
- Add standard form fields e.g. Author, Dates, Access Level, State to the add / edit form view
- Workflows marked default must not be deleted
- An unpublished workflow must not be selected as the default (see bug list)
- Help button missing in both list view and add / edit workflow form
- Description / Editor field shows other different toolbar for add / edit workflow form than add / edit article
- New workflow status (required) is "Trashed". It should be published or at least unblished.
- The title of the first (and only) "Tab" in the add / edit view reads "Basis" where it should be "New / Edit Workflow" depending on the context
- Language strings e.g.: COM_WORKFLOW_ARE_YOU_SURE
- Created by (author) can not be set (bug or feature)
- Author filter in the table tools does not work
- The "Default" toggle button in the workflow list for a workflow shows the following success messages: "Item set to home" and "Item uset home"
	RFC: Even with messages like: "The selected workflow has been set as the default." or "Set as default", I'm not sure if the opposite really should be "Unset default", which is the current tooltip.
	As said, we have a switch / toggle behavior now, that allows us to "unset" all workflows. If I remember correctly, at least one workflow must be
	set as the systems default.
- The currect order and limit fields of the workflows list view "jump" in size onchange / page refresh
- If an author gets deleted, the field should say: N/A (it is currently empty)
- Trashed success message says: COM_WORKFLOW_N_ITEMS_TRASHED
- Set workflow to default while another is already default shows the following error message:
	Default workflow already is
	Save failed with the following error:

------------------------------------

Workflow(s) RFC / TBD
- Remove the "transitions" link in the list and rename the status link to "manage" (Benjamin Trenkle)
- Remove "modification date" from the workflows view (but should remain in the db/form) (Benjamin Trenkle)
	- I'm not sure. Personally, I would be interested to sort by the last modified workflow, maybe even in combination with an author. I defenetly
	would want to have this in the add / edit form. (Sven Hurt)

------------------------------------

Transitions:
- Clicking on "Show" (Transition) in the list of worklows shows an empty error message
- Transition can not be added:
	Error
	Save failed with the following error:
- What use case is behind that "From" - "To" State filter? "List all Transitions between two states" - what does one want with that list?
- The "From" - "To" filter does not list existing states regardless of worklow and / or published state
- Saving a new transition with the same from / to states gives the user -> two <- error messages:
	"You choose the same state from and to" and "Save failed with the following error:"
	While is is correct to not be able to save a transition that way, the error message itself should be more clear.
- List view can not be ordered by "Status"

------------------------------------

States:
- Toggle "Default" in states list view leads to blank page
- Table / List ordering shows "by Status", but the list doesn't provide that field
- Two+ State entries can be set to "Default"
- After saving a new State, the sub navigation looses all entries but States and Transitions
- States are saved with the access field set to 0 in the db
- States can deleted directly wihout being set to trashed. See general info / task. It should be impossible to delete or unpublish a transition
	which is "in use".
- At least one default item must be set (currect situation allows to toggle all)
- The main menu closes when states was selected in the list of workflows
- Two states set as default leads to an exception: Class 'Joomla\Component\Workflow\Administrator\Model\Factory' not found.
- Sub-menu missing in states view
- Is there a reason why states can not be unpublished / trashed? From what I found in the chat, there was not a clear reason of why this has been left out, correct? -> POLS <-

------------------------------------

Usability / Workflow (RFC/TBD):
- Differentiation between extension types and / or demarcation to other extensions (link in every extension ~> redundancy)
- Current implementation allows to open the transitions list / edit view without having set a states to choose from first
- States / Tranitions link on top of the Articles link (confusing / breaking context)
- Correlation between Workflow - Transition and State calls for a "back" button

------------------------------------

"Un-/Related" Joomla! 4.x Bugs (RFC/TBD ~ potential blockers in general):
- Status in add / edit article form not mandatory / required
- Readonly fields seem to be validated
- Article cannot be saved:
	An error has occurred.
		1054 Unknown column 'from_status_id' in 'where clause'
- Search & Filter field doesn't work (choosen + JS errors)
- Search Tools Toggle does not change icon (arrow)
- List error / success messages cover action button
- Default ordering "ID ascending" missing arrow indicator
- Too much space between label and input element of entity titles
- Description text of acl actions are not interpreted as html. See articles permissions tab and hover over "Delete"

------------------------------------

General (RFC/TBD):
- Rename com_workflow to com_workflows
- Rename list view title from "Workflows List" to "Workflows" (same for Transitions and States)
- Checkout functionality missing (checked_out, checked_out_time)
- Created and modified datetime fields only contain the date, not the time
- ACL implementation
- Forms directory / method seems to not correct (forms vs Model vs models/forms)
- Incorrect colspan count in "all" list views
- The ordering of the database fields for workflows, transitions and states should be equal to other core tables
- Generaly every entity but a user, must be set to "Tashed" before it can be deleted (fail-safe). The Workflow component
	should adhere to that principle for workflows, transitions and states alike.

------------------------------------

Category:
- The category workflow field should offer a "Use default" selection + a list of existing workflows. New articles created in that category should be assigned to the selected workflow automatically.
	- The core / default workflow do be added (installation + sample data) should reflect the current Joomla! 3.x behavior e.g.:
    	States: Published, Unpublished, Archived, Trashed
		Transitions: P => U/A/T, U => P/T, A => T/P, T => P/U
		Note: The workflow field should only list workflows that are published and match the correct access level
		- The check for that must be included in the MVC structure! (Prevent request manipulation)
		- The value for the default workflow in the selection field should be an empty string, not an / the ID
	- The title / caption of the workflow tab in the category add / edit view form should be "Workflow" (singular), where the field label should read "Workflows"

------------------------------------

Content

	Convert the status dropdown (This too i don't fully understand :)) Janek
    	At the moment there is the default status dropdown in the form view, this should handle the new system Benjamin

    Convert the publishing icons in the list (To what icon i should change publishing icons?) Janek
		They should be the same, but use the conditions of the current status Benjamin

    Add toolbar buttons (Could you explain? To run more then one transition?) Janek
		Yes, to run transitions for more articles, but is a "nice to have". (Make sure, to remov the current "publish/unpublish" buttons Benjamin

**Comment by shubhamnba2009** [comment]
https://github.com/joomla/joomla-cms/issues/8421
I came around one of this closed issue that says it is PHP version responsible for the clear button bug in search tools filter. Is it possible that maybe it is responsible here too!!

**Comment by jjaracz** [comment]
Workflow(s):

> - Add search / filter to list view
> 	Hint: Either global layout or custom layout depending on flexibility
> 		Possible Fields: [Search][Clear][Search Tools*][Status][Default (RFC / TBD][Author]
> 		Note: The ordering of those fields should reflect the order of the table headings (see other core components)
> 	Dependencies: Model, View (View Layouts), Global Layouts
> 	There are at least two exsting issues / tickets regarding the Search Tools Layout. One is about a bug that already has an influence on the
> 	list views "Clear" button for the search field. The second ist about a refactoring of the Search Tools in general.
> 	Attention: Both could become a blocker for this task! The same goes for all other filters including: Transitions and States!

That task / issue / ticket i don't understand at all ...

> 		Note: The wording AND ordering should be the same across all extensions e.g. created fields (by vs at vs date)

Could you explain?

> - Add publish, unpublish and trash toggle functionality the same way it is implemented in other core extensions

That i think is already.

- add access level checks for buttons in toolbar.

> - Add the following toolbar buttons to the add / edit form view: Save / Save & Close / Save & New

What you mean by that? I added that at the beginning of project.

> - Add standard form fields e.g. Author, Access Level to the add / edit form view

We should have access level for workflows or only for transitions?

> - An unpublished workflow must not be selected as the default (see bug list)

I think that was sloved in some commits before

> - Description / Editor field shows other different toolbar for add / edit workflow form than add / edit article

Because in article we can attach additional elements like images, modules and something like that. I do not think that is necessary for our use case.

> - New workflow status (required) is "Trashed". It should be published or at least unblished

Could you explain? For default there is published when you try to add new workflow, if that is what you mean :)

> - Language strings e.g.: COM_WORKFLOW_ARE_YOU_SURE

I do not get it.

> - Created by (author) can not be set (bug or feature)

I think that is feature. For me author should be guy who create item.

> - The "Default" toggle button in the workflow list for a workflow shows the following success messages: "Item set to home" and "Item uset home"
> 	RFC: Even with messages like: "The selected workflow has been set as the default." or "Set as default", I'm not sure if the opposite really should be "Unset default", which is the current tooltip.
> 	As said, we have a switch / toggle behavior now, that allows us to "unset" all workflows. If I remember correctly, at least one workflow must be
> 	set as the systems default.

We can set default workflow and that mean that workflow will be for whole component, we not have to set default workflow because when there will be no default then still categories are set for workflows.

> - The currect order and limit fields of the workflows list view "jump" in size onchange / page refresh 
You mean left side panel is opening and closing?

> - Set workflow to default while another is already default shows the following error message:
> 	Default workflow already is
> 	Save failed with the following error:

Yep, I know i was trying figure out why this is happening but i doubted ...

------------------------------------

Workflow(s) RFC / TBD

> - Remove the "transitions" link in the list and rename the status link to "manage" (Benjamin Trenkle)

So, where will be redirection to transitions view (expect left side panel with a lot of links)?

- Remove "modification date" from the workflows view (but should remain in the db/form) (Benjamin Trenkle)
	- I'm not sure. Personally, I would be interested to sort by the last modified workflow, maybe even in combination with an author. I defenetly
	would want to have this in the add / edit form. (Sven Hurt)

   - I agree with Sven (author already is in list ;) ) (Janek)

**Comment by bembelimen** [comment]
> > Workflow(s):
> > 		Note: The wording AND ordering should be the same across all extensions e.g. created fields (by vs at vs date)
> 
> Could you explain?

The default field name is "Created By" and "Create Date" instead of "Created at" (=> see com_content/com_categories), so you should use the same naming (also the same name for the database fields) 

> > - Add standard form fields e.g. Author, Access Level to the add / edit form view
> 
> We should have access level for workflows or only for transitions?

Access level is not needed here

> > - New workflow status (required) is "Trashed". It should be published or at least unblished
> 
> Could you explain? For default there is published when you try to add new workflow, if that is what you mean :)

Yep, it's fixed

> > - Language strings e.g.: COM_WORKFLOW_ARE_YOU_SURE
> 
> I do not get it.

Not all language strings are translated (like if you trash a workflow)

> > - Created by (author) can not be set (bug or feature)
> 
> I think that is feature. For me author should be guy who create item.

In Joomla! a user with specific rights (core.admin) can set other creators (e.g. when the original creator leaves the page and will be deleted)

> > - The "Default" toggle button in the workflow list for a workflow shows the following success messages: "Item set to home" and "Item uset home"
> > 	RFC: Even with messages like: "The selected workflow has been set as the default." or "Set as default", I'm not sure if the opposite really should be "Unset default", which is the current tooltip.
> > 	As said, we have a switch / toggle behavior now, that allows us to "unset" all workflows. If I remember correctly, at least one workflow must be
> > 	set as the systems default.
> 
> We can set default workflow and that mean that workflow will be for whole component, we not have to set default workflow because when there will be no default then still categories are set for workflows.

The translation is wrong: "Item set to home".

> > - The currect order and limit fields of the workflows list view "jump" in size onchange / page refresh 
> You mean left side panel is opening and closing?

Sort by "ID ASC"
![grafik](https://user-images.githubusercontent.com/1229869/28442264-cb3974bc-6daf-11e7-8465-a575e6d10309.png)

Sort by "ID DESC"
![grafik](https://user-images.githubusercontent.com/1229869/28442278-e43bf282-6daf-11e7-857e-aa86e37f7e92.png)

> > - Set workflow to default while another is already default shows the following error message:
> > 	Default workflow already is
> > 	Save failed with the following error:
> 
> Yep, I know i was trying figure out why this is happening but i doubted ...

We'll look into it.

> > - Remove the "transitions" link in the list and rename the status link to "manage" (Benjamin Trenkle)
> 
> So, where will be redirection to transitions view (expect left side panel with a lot of links)?

The links on the left side will be the same (one status, one transitions)

> > - Remove "modification date" from the workflows view (but should remain in the db/form) (Benjamin Trenkle)
> > 	- I'm not sure. Personally, I would be interested to sort by the last modified workflow, maybe even in combination with an author. I defenetly
> > 	would want to have this in the add / edit form. (Sven Hurt)
> > 
>    - I agree with Sven (author already is in list ;) ) (Janek) 

OK for me, but with the correct naming)

**Comment by jjaracz** [comment]
Workflow(s):

> 
> > > - Created by (author) can not be set (bug or feature)
> > 
> > I think that is feature. For me author should be guy who create item.
> 
> In Joomla! a user with specific rights (core.admin) can set other creators (e.g. when the original creator leaves the page and will be deleted)

So then i will add option to set author for the workflow :)

> > - The currect order and limit fields of the workflows list view "jump" in size onchange / page refresh 
> You mean left side panel is opening and closing?

> Sort by "ID ASC"
> ![grafik](https://user-images.githubusercontent.com/1229869/28442264-cb3974bc-6daf-11e7-8465-a575e6d10309.png)
> 
> Sort by "ID DESC"
> ![grafik](https://user-images.githubusercontent.com/1229869/28442278-e43bf282-6daf-11e7-857e-aa86e37f7e92.png)

Do you know why this happen? I was looking for some error in the code but i do not find :/

> > > - Set workflow to default while another is already default shows the following error message:
> > > 	Default workflow already is
> > > 	Save failed with the following error:
> > 
> > Yep, I know i was trying figure out why this is happening but i doubted ...
> 
> We'll look into it.

Great.

**Comment by matrikular** [comment]
> Add search / filter to list view
> Hint: Either global layout or custom layout depending on flexibility
> Possible Fields: [Search][Clear][Search Tools*][Status][Default (RFC / TBD][Author]
> Note: The ordering of those fields should reflect the order of the table headings (see other core components)
> Dependencies: Model, View (View Layouts), Global Layouts
> There are at least two exsting issues / tickets regarding the Search Tools Layout. One is about a bug that already has an influence on the
> list views "Clear" button for the search field. The second ist about a refactoring of the Search Tools in general.
> Attention: Both could become a blocker for this task! The same goes for all other filters including: Transitions and States!

**Answer / Explanation:** Joomla! uses layouts for reusable markup. These layouts can be overriden on a extension basis. If your toolbar, filter or table layout is somehow different (e.g. has custom fields) from others, you would have to create such a layout override.

Your task is to evaluate what fields are needed for each list view, in this case, the workflow (but not limited to). At the time of preparing this list, the extension used the old search toolbar and we were discussing the "clear search" bug. That is why I included the "warning". If you encounter more bugs here, estimate the impact of it:
- Is someone already working on that?
- Can I continue without the functionality?
- Can I help fix to problem so I can continue?

Now that you switched to the table options, you have to evaluate if all required fields are A. present B. in order and C. that they work as expected.

------------------------------------------------------------------------

> Note: The wording AND ordering should be the same across all extensions e.g. created fields (by vs at vs date)

**Answer / Explanation:** Look at the views / layouts of other core extensions. What is the common wording and ordering of the fields used there? Check if the table structure is in line with other core extension tables (field names, ordering, indices)?

**Note:** See / Use comments like this as a checklist. If everything is fine AND working, then there is nothing more to do and you can move on to the next task.

------------------------------------------------------------------------

> Add publish, unpublish and trash toggle functionality the same way it is implemented in other core extensions

**Answer / Explanation:** Again, a checklist ... see if that is really the case for all your views - if so, move on.

------------------------------------------------------------------------

> Add standard form fields e.g. Author, Access Level to the add / edit form view

**Answer / Explanation:** This is an older task / entry and can be ignored for now. Again, this is more about unification.  Since we have an asset_id field in the table, the last step for com_workflow(s) would be to implement the ACL (permissions), but after talking to Benjamin, that is a task for us later on, so don't worry about that for now.

------------------------------------------------------------------------

> Add the following toolbar buttons to the add / edit form view: Save / Save & Close / Save & New

**Answer / Explanation:** We currently have Save / Save & Close and Cancel across all entities. Is this the final list of action buttons to use for now and / or do you have plans to add maybe Save & New or a Help button? Keep in mind that the wording of the Cancel button changes in regards to the current state (new vs edit).

------------------------------------------------------------------------

> An unpublished workflow must not be selected as the default (see bug list)

**Answer / Explanation:**
Test: Set a workflow as default in the list view, now unblish it either in the list view or edit view.
Result: The workflow can be unpublished and save as the default.
Expected: Either the un-publish state change or the set to default action should be prevented.

------------------------------------------------------------------------

> Add standard form fields e.g. Author, Access Level to the add / edit form view

**Answer / Explanation:** e.g => exempli gratia = just an example. Align the views / layouts to other core extensions and add the fields that are required / needed.

------------------------------------------------------------------------

> Language strings e.g.: COM_WORKFLOW_ARE_YOU_SURE

**Answer / Explanation:** Make sure that all language strings are translated and in line with the core. I wasn't sure if "_ARE_YOU_SURE" is a term we use in the project.

------------------------------------------------------------------------

> Created by (author) can not be set (bug or feature)

**Answer / Explanation:** As you know, we don't have referential integrity in Joomla!. If an author gets deleted, the entity is ownerless and that at the very least will be reflected in the list view.

No assigned author => N/A in list view.

------------------------------------------------------------------------

> The "Default" toggle button in the workflow list for a workflow shows the following success messages: "Item set to home" and "Item uset home"
> RFC: Even with messages like: "The selected workflow has been set as the default." or "Set as default", I'm not sure if the opposite really should be "Unset default", which is the current tooltip.
> 
> As said, we have a switch / toggle behavior now, that allows us to "unset" all workflows. If I remember correctly, at least one workflow must be set as the systems default.

**Answer / Explanation:** We don't set anything to "home" here and "uset" is not a word. A cosmetic bug that can be fixed easily and crossed from the list. The bigger task here though, is that you can toggle the default state in the list.

> We can set default workflow and that mean that workflow will be for whole component, we not have to set default workflow because when there will be no default then still categories are set for workflows.

**Answer / Explanation:** That is not correct. We MUST have  one (1) workflow set as default. For all new installations, this workflow is represented by the set of transitions and states that describe the current "workflow" for articles in Joomla! < 4 (see what Benjamin wrote). If you don't have a default workflow then the extension or category selection would fall back to the first entry in the list of available workflows.

**In short:** A user can set a workflow to default in the list, but he / she must not be able to "toggle" so that no workflow is selected.

**Comment by jjaracz** [comment]
Transitions:

> - Clicking on "Show" (Transition) in the list of worklows shows an empty error message

I deleted show button.

> - Transition can not be added:
> 	Error
> 	Save failed with the following error:

Maybe that is issue from some previous commits because now it is working :)

> - What use case is behind that "From" - "To" State filter? "List all Transitions between two states" - what does one want with that list?

I do not get it. Filter "From" and "To" is important because we can see how many we created transitions with the same e.g. starting state. Where is that string from second question?

> - The "From" - "To" filter does not list existing states regardless of worklow and / or published state

Yes, i publish message on glip, maybe someone of you would know what is going on :)

> - Saving a new transition with the same from / to states gives the user -> two <- error messages:
> 	"You choose the same state from and to" and "Save failed with the following error:"
> 	While is is correct to not be able to save a transition that way, the error message itself should be more clear.

Yep, that is the issue what i mention before, Benjamin will look into it ;)

**Comment by matrikular** [comment]
> Clicking on "Show" (Transition) in the list of worklows shows an empty error message
>> I deleted show button.

I have a bad feeling about this. We should schedule a meeting for this to be discussed in detail, please. Right now, the way the user manages workflows, transitions and states seems hard to grasp / unintuitive. Removing the link to the transitions made it somehow worse. Will you @jjaracz  or @bembelimen set up a date and time maybe?

> Transition can not be added:
> Error
> Save failed with the following error:
>>Maybe that is issue from some previous commits because now it is working :)

I have the latest updates from the 0.0.1 branch set up and called:
`administrator/index.php?option=com_workflow&view=transitions&workflow_id=1&extension=com_content`

Here I clicked on "New" and entered "My Second Transition". After hitting the save button, the same error message as above popped up.

> What use case is behind that "From" - "To" State filter? "List all Transitions between two states" - what does one want with that list?
>>I do not get it. Filter "From" and "To" is important because we can see how many we created transitions with the same e.g. starting state. Where is that string from second question?

Okay, I hope to be able to see the bigger picture when the default workflow and its transitions / states are set up.

**Comment by jjaracz** [comment]
> > Clicking on "Show" (Transition) in the list of worklows shows an empty error message
> >> I deleted show button.
> 
> I have a bad feeling about this. We should schedule a meeting for this to be discussed in detail, please. Right now, the way the user manages workflows, transitions and states seems hard to grasp / unintuitive. Removing the link to the transitions made it somehow worse. Will you @jjaracz  or @bembelimen set up a date and time maybe?

Sure, on the glip we can agree meeting :)

> > Transition can not be added:
> > Error
> > Save failed with the following error:
> >>Maybe that is issue from some previous commits because now it is working :)
> 
> I have the latest updates from the 0.0.1 branch set up and called:
> `administrator/index.php?option=com_workflow&view=transitions&workflow_id=1&extension=com_content`

Try now, there was a small issue with filtering and now should be resolved, but strange is that, that error shows on adding view ... Maybe you have not set states? Only things that are required to save transition are title, states and ofc status.

**Comment by jjaracz** [comment]
States:
> - The main menu closes when states was selected in the list of workflows

I do no see that. Could you take a screenshot?

> - Sub-menu missing in states view

Where?

> - Is there a reason why states can not be unpublished / trashed? From what I found in the chat, there was not a clear reason of why this has been left out, correct? -> POLS <-

Yep, you are right, i added that in the near commit :)

**Comment by jjaracz** [comment]
@bembelimen
